server/src/server/kb/concept/TypeConverter.java

```diff
@@ -0,0 +1,133 @@
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package grakn.core.server.kb.concept;
+
+import com.google.common.collect.ImmutableMap;
+import grakn.core.concept.type.AttributeType;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.Map;
+
+public abstract class TypeConverter<DESERIALISED>  extends Serialiser<DESERIALISED, Object> {
+
+    private static Map<AttributeType.DataType<?>, TypeConverter<?>> converters = ImmutableMap.<AttributeType.DataType<?>, TypeConverter<?>>builder()
+            .put(AttributeType.DataType.BOOLEAN, new Default<Boolean>())
+            .put(AttributeType.DataType.DATE, new DateConverter())
+            .put(AttributeType.DataType.DOUBLE, new DoubleConverter())
+            .put(AttributeType.DataType.FLOAT, new FloatConverter())
+            .put(AttributeType.DataType.INTEGER, new IntegerConverter())
+            .put(AttributeType.DataType.LONG, new LongConverter())
+            .put(AttributeType.DataType.STRING, new Default<String>())
+            .build();
+
+
+    public static <DESERIALISED> TypeConverter<DESERIALISED> of(AttributeType.DataType<DESERIALISED> dataType) {
+        TypeConverter<?> converter = converters.get(dataType);
+        if (converter == null){
+            throw new UnsupportedOperationException("Unsupported DataType: " + dataType.toString());
+        }
+        return (TypeConverter<DESERIALISED>) converter;
+
+    }
+
+    public static class Default<DESERIALISED> extends TypeConverter<DESERIALISED>{
+        @Override
+        public Object serialise(DESERIALISED value) { return value; }
+
+        @Override
+        public DESERIALISED deserialise(Object value) { return (DESERIALISED) value;}
+    }
+
+    public static class DateConverter extends TypeConverter<LocalDateTime> {
+        @Override
+        public Object serialise(LocalDateTime value) {
+            return value.atZone(ZoneId.of("Z")).toInstant().toEpochMilli();
+        }
+
+        @Override
+        public LocalDateTime deserialise(Object value) {
+            if (value instanceof LocalDateTime) return (LocalDateTime) value;
+            if (value instanceof Long) return LocalDateTime.ofInstant(Instant.ofEpochMilli((Long) value), ZoneId.of("Z"));
+            throw new ClassCastException();
+        }
+    }
+
+    public static class DoubleConverter extends TypeConverter<Double> {
+        @Override
+        public Double deserialise(Object value) {
+            if (value instanceof Number){
```

Comment:
what happens when you cast a `Long` to a `Double` or `Float`?

Reply:
I don't get the question. If it's a Number (so Long as well) it gets converted to a Double.